File: Data Structures & Algorithms/coin-change/submission-2.cpp
class Solution {

public:
    int coinChange(vector<int>& coins, int amount) {
        int n=coins.size();
        vector<int>prev(amount+1,0);
        vector<int>cur(amount+1,0);
        for(int i=0;i<=amount;i++){
            if(i%coins[0]==0)prev[i]=i/coins[0];
            else prev[i]=1e9;
        }

        for(int i=1;i<n;i++){
            for(int target=0;target<=amount;target++){
                int nottake=prev[target];
                int take=1e9;
                if(target>=coins[i])take=1+cur[target-coins[i]];
                cur[target]=min(nottake,take);
            }
            prev=cur;
        }
        return (prev[amount]>=1e9)?-1:prev[amount];
    }
};

File: Data Structures & Algorithms/target-sum/submission-2.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        n=len(nums)
        dp = {}

        def fun(i,cur):
            if(i==n and cur==0):
                return 1

            if(i==n):
                return 0

            if(i,cur) in dp:
                return dp[(i,cur)]

            pos=fun(i+1,cur-nums[i])
            neg=fun(i+1,cur+nums[i])

            dp[(i,cur)]= pos+neg
            return dp[(i,cur)]


        return fun(0,target)